sysl2/sysl/deps.go

```diff
@@ -0,0 +1,264 @@
+			for _, choice := range c.Alt.GetChoice() {
+				cs.CollectCallStatements(choice.GetStmt())
+			}
+		default:
+			panic("No statement!")
+		}
+	}
+}
+
+func FindApps(module *sysl.Module, excludes, integrations StrSet, ds []*AppDependency, highlight bool) StrSet {
+	r := []string{}
+	appReStr := toPattern(integrations.ToSlice())
+	re := regexp.MustCompile(appReStr)
+	st, _ := stream.Of(ds)
+	st.ForEach(func(v stream.T) bool {
+		if dep, ok := v.(*AppDependency); ok {
+			appNames := dep.extractAppNames()
+			excludeApps := appNames.Intersection(excludes)
+			if len(excludeApps) > 0 {
+				return true
+			}
+			highlightApps := appNames.Intersection(integrations)
+			if !highlight && len(highlightApps) == 0 {
+				return true
+			}
+			appStream, _ := stream.Of(appNames.ToSlice())
+			appStream.ForEach(func(v stream.T) bool {
+				item := v.(string)
+				app := module.GetApps()[item]
+				if highlight {
+					if re.MatchString(item) &&
+						!hasPattern("human", app.GetAttrs()) {
+						r = append(r, item)
+					}
+					return true
+				}
+				if !hasPattern("human", app.GetAttrs()) {
+					r = append(r, item)
+				}
+				return true
+			})
+		}
+		return true
+	})
+
+	return MakeStrSet(r...)
+}
+
+func walkPassthrough(excludes, passthroughs StrSet, dep *AppDependency, integrations *DependencySet, module *sysl.Module) {
+	target := dep.Target
+	targetName := target.Name
+	targetEndpoint := target.Endpoint
+	excludedApps := MakeStrSet(targetName).Intersection(excludes)
+	undeterminedEps := MakeStrSet(targetEndpoint).Intersection(MakeStrSet(".. * <- *", "*"))
+	//Add to integration array since all dependencies are determined.
+	if len(excludedApps) == 0 && len(undeterminedEps) == 0 {
+		integrations.Deps[dep.String()] = dep
+	}
+
+	// find the next outbound dep
+	if passthroughs.Contains(targetName) {
+		cs := NewCallStatementSlice()
+		cs.CollectCallStatements(module.GetApps()[targetName].GetEndpoints()[targetEndpoint].GetStmt())
+		for _, stmt := range cs.GetSlice() {
+			call := stmt.GetStmt().(*sysl.Statement_Call).Call
+			nextAppName := strings.Join(call.GetTarget().GetPart(), " :: ")
+			nextEpName := call.GetEndpoint()
+			next := MakeAppElement(nextAppName, nextEpName)
+			nextDep := MakeAppDependency(target, next, stmt)
+			walkPassthrough(excludes, passthroughs, nextDep, integrations, module)
+		}
+	}
+}
+
+func (ds *DependencySet) FindIntegrations(apps, excludes, passthroughs StrSet, module *sysl.Module) *DependencySet {
+	integrations := NewDependencySet()
+	outboundDeps := NewDependencySet()
+	lenPassthroughs := len(passthroughs)
+	for _, dep := range ds.Deps {
+		appNames := dep.extractAppNames()
+		endpoints := dep.extractEndpoints()
+		isSubsection := isSub(appNames, apps)
+		isSelfSubsection := isSub(MakeStrSet(dep.Self.Name), apps)
+		isTargetSubsection := isSub(MakeStrSet(dep.Target.Name), passthroughs)
+		interExcludes := appNames.Intersection(excludes)
+		interEndpoints := endpoints.Intersection(MakeStrSet(".. * <- *", "*"))
```

Comment:
`MakeStrSet(".. * <- *", "*")` take this out of the loop.

Reply:
yep. moved it out. it should be created only once.